pkg/controllermanager/controller/cloudprofile/cloudprofile_control.go

```diff
@@ -111,13 +111,53 @@ func (r *cloudProfileReconciler) Reconcile(ctx context.Context, request reconcil
 		cloudProfileLogger.Info(message)
 		r.recorder.Event(cloudProfile, corev1.EventTypeNormal, v1beta1constants.EventResourceReferenced, message)
 
-		return reconcile.Result{}, fmt.Errorf("CloudProfile %q still has references", cloudProfile.Name)
+		return reconcile.Result{}, fmt.Errorf("Cannot delete CloudProfile %q, because the following Shoots are still referencing it: %+v", cloudProfile.Name, associatedShoots)
 	}
 
 	if err := controllerutils.PatchAddFinalizers(ctx, r.gardenClient, cloudProfile, gardencorev1beta1.GardenerName); err != nil {
-		logger.Logger.Errorf("could not add finalizer to CloudProfile: %s", err.Error())
+		cloudProfileLogger.Errorf("could not add finalizer to CloudProfile: %s", err.Error())
 		return reconcile.Result{}, err
 	}
 
+	// TODO voelzmo - this migration step ensures that all MachineImageVersions in the Cloud Profile contain `docker` in their list of supported Container Runtimes
+	// This can be removed in a couple of versions. Note that while this is still in here, it is impossible to add an image without `docker` support!
+	migrationHappened := migrateMachineImageVersionCRISupport(cloudProfile)
+
+	if migrationHappened {
+		cloudProfileLogger.Infof("migrated Machine Image Versions to explicitly contain `docker` as supported CRI")
+		if err := r.gardenClient.Update(ctx, cloudProfile); err != nil {
+			cloudProfileLogger.Errorf("failed to update Cloud Profile spec: %+v", err)
+			return reconcile.Result{}, err
+		}
+	}
+
 	return reconcile.Result{}, nil
 }
+
+func migrateMachineImageVersionCRISupport(cloudProfile *gardencorev1beta1.CloudProfile) bool {
+	var migrationHappened bool
+	for i, image := range cloudProfile.Spec.MachineImages {
+		for j, version := range image.Versions {
+			if version.CRI == nil {
+				cloudProfile.Spec.MachineImages[i].Versions[j].CRI = []gardencorev1beta1.CRI{{Name: gardencorev1beta1.CRINameDocker}}
+				migrationHappened = true
+				continue
+			}
```

Comment:
Nit: This block is not needed.

Reply:
Good catch! How did this end up in here? ;o